rust/lance/src/encodings/binary.rs

```diff
@@ -251,42 +261,54 @@ impl<'a, T: ByteArrayType> BinaryDecoder<'a, T> {
 
         Ok(Arc::new(GenericByteArray::<T>::from(array_data)))
     }
+}
 
-    async fn take_internal(
-        &self,
-        positions: &Int64Array,
-        indices: &UInt32Array,
-    ) -> Result<ArrayRef> {
-        let start = indices.value(0);
-        let end = indices.value(indices.len() - 1);
-        let array = self
-            .get_range(positions, start as usize..end as usize + 1)
-            .await?;
-        let adjusted_offsets = subtract_scalar(indices, start)?;
-        Ok(take(&array, &adjusted_offsets, None)?)
-    }
+#[derive(Debug)]
+struct TakeChunksPlan {
+    indices: UInt32Array,
+    is_contiguous: bool,
 }
 
+/// Group the indices into chunks, such that either:
+/// 1. the indices are contiguous (and non-repeating)
+/// 2. the values are within `min_io_size` of each other (and thus are worth
+///    grabbing in a single request)
 fn plan_take_chunks(
     positions: &Int64Array,
     indices: &UInt32Array,
     min_io_size: i64,
-) -> Result<Vec<UInt32Array>> {
+) -> Result<Vec<TakeChunksPlan>> {
```

Comment:
do we need to find another plan if `len(indices)` is small? say it is just 2 indices per data file.

Reply:
Not sure. I think we should do another round of optimizations for random access later, as I suspect there are improvements to be made there. I just focused on the filter case, since that had the worst performance.